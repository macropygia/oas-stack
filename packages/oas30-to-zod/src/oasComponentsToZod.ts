import path from 'node:path'
// import { fileURLToPath } from 'node:url'

import ejs from 'ejs'
import fse from 'fs-extra'
import SwaggerParser from '@apidevtools/swagger-parser'
import { red } from 'ansis/colors'
import type { OpenAPI, OpenAPIV3 } from 'openapi-types'

import { defaultParseOptions, parseSchema } from '@/parseSchema.js'
import type {
  Document,
  ComponentGraph,
  Options,
  ComponentName,
} from '@/types/index.js'
import { format } from '@/utils/format.js'
import { generatePrintingOrder } from '@/utils/generatePrintingOrder.js'
import { autocompleteObject } from '@/utils/autocompleteObject.js'
import { isV3, isValidDoc } from '@/utils/validateDoc.js'

// const __filename = fileURLToPath(import.meta.url)
// const __dirname = path.dirname(__filename)

const defaultOptions: Options = {
  output: false,
  dereference: false,
  disableAutocomplete: false,
  exportName: 'schemas',
  individually: false,
  eslintDisable: false,
  disableRules: [],
  withoutImport: false,
  withoutExport: false,
  inheritPrettier: false,
  disableFormat: false,
  ...defaultParseOptions,
}

/**
 * Generate Zod schemas from OpenAPI 3.0 Components Object
 * @param input - Document object or file path
 * @param userOptions - Options
 * @returns Zod schemas (JavaScript)
 */
export const oasComponentsToZod = async (
  input: OpenAPIV3.Document | string,
  userOptions?: Options
): Promise<string> => {
  const options: Options = { ...defaultOptions, ...userOptions }

  const doc = await init(input, options)

  const dereferencedDoc = await SwaggerParser.dereference(
    JSON.parse(JSON.stringify(doc))
  ).then((deref) => deref as Document)

  const graph = generateGraph(dereferencedDoc)

  const targetDoc = options.dereference ? dereferencedDoc : doc
  const targetSchema = targetDoc.components.schemas

  // Reserved object for cusomization
  const data: Record<string, any> = {}

  const parsedComponents: Record<ComponentName, string> = {}
  Object.entries(targetSchema).forEach(([compName, compSchema]) => {
    const deps: ComponentName[] = []
    parsedComponents[compName] = parseSchema(compSchema, {
      options,
      parsers: options.parsers,
      graph,
      name: compName,
      deps,
      doc: targetDoc,
      schemas: targetDoc.components.schemas,
      data,
    })
    graph.deps[compName] = deps
  })

  // Create a shallow copy of deps for destructive processing
  const componentDeps = Object.assign({}, graph.deps)
  const printingOrder = generatePrintingOrder(componentDeps)

  const template = options.template
    ? fse.readFileSync(options.template).toString()
    : fse.readFileSync(path.resolve(__dirname, 'default.ts.ejs')).toString()

  const rawParsed = ejs.render(template, {
    // Processed data
    parsedComponents,
    printingOrder,
    // Options
    ...options,
    // Graph
    graph,
    data,
  })

  const parsed = options.disableFormat
    ? rawParsed
    : format(rawParsed, options.inheritPrettier)

  if (options.output) fse.outputFileSync(options.output, parsed)

  return parsed
}

/**
 * Verify settings and document
 * @param input - Document object or file path
 * @param options - Options
 * @returns Document object
 */
const init = async (input: OpenAPI.Document | string, options: Options) => {
  if (options.output)
    await fse.ensureFile(options.output).catch((err) => {
      console.error(red`Unable to create output file.`)
      throw new Error(err)
    })

  if (options.template && !fse.existsSync(options.template))
    throw new Error(`EJS template does not exists: ${options.template}`)

  const doc =
    typeof input === 'string' ? await SwaggerParser.bundle(input) : input

  if (!isV3(doc)) {
    console.error(
      red`Document version mismatch. Only version 3.0 is supported.`
    )
    throw new Error('Only version 3.0 is supported.')
  }

  if (!isValidDoc(doc))
    throw new Error(`Document has no 'components' or 'schemas'.`)

  if (!options.disableAutocomplete) autocompleteObject(doc)

  return doc
}

/**
 * Generate component graph from dereferenced document
 * @param dereferencedDoc - Dereferenced document
 */
const generateGraph = (dereferencedDoc: Document): ComponentGraph => {
  const compGraph: ComponentGraph = {
    deps: {},
    isObject: {},
    isNullable: {},
    hasDefault: {},
  }

  const resolvedSchemas = (dereferencedDoc as Document).components.schemas

  Object.entries(resolvedSchemas).forEach(([compName, compSchema]) => {
    // Object
    if ('type' in compSchema && compSchema.type === 'object')
      compGraph.isObject[compName] = true
    else compGraph.isObject[compName] = false

    // Nullable
    if ('nullable' in compSchema && compSchema.nullable === true)
      compGraph.isNullable[compName] = true
    else compGraph.isNullable[compName] = false

    // Default
    if ('default' in compSchema) compGraph.hasDefault[compName] = true
    else compGraph.hasDefault[compName] = false
  })

  return compGraph
}
