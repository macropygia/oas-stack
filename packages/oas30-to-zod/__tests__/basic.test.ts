import fse from 'fs-extra'
import { describe, test, expect } from 'vitest'

import { oasComponentsToZod } from '../src/oasComponentsToZod.js'

const minimumOutput = {
  withoutImport: true,
  withoutExport: true,
}

describe('oasComponentsToZod (basic)', async () => {
  test('Default', async () => {
    await expect(
      oasComponentsToZod('__tests__/basic.yml')
    ).resolves.toMatchSnapshot()
  })

  test('Disable format', async () => {
    await expect(
      oasComponentsToZod('__tests__/basic.yml', {
        disableFormat: true,
      })
    ).resolves.toMatchSnapshot()
  })

  test('Enable output', async () => {
    await oasComponentsToZod('__tests__/minimum.yml', {
      ...minimumOutput,
      output: '__tests__/minimum.ts',
    })

    expect(fse.readFileSync('__tests__/minimum.ts').toString())
      .toMatchInlineSnapshot(`
      "const Minimum = z.string();
      "
    `)

    fse.removeSync('__tests__/minimum.ts')
  })

  test('Enable dereference', async () => {
    await expect(
      oasComponentsToZod('__tests__/dereference.yml', {
        ...minimumOutput,
        dereference: true,
      })
    ).resolves.toMatchSnapshot()
  })

  test('Change export variable name', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        withoutImport: true,
        exportName: 'foo',
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z.string();

      export const foo = {
        Minimum,
      };
      "
    `)
  })

  test('Export individually', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        individually: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "export const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable`', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable */

      const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable` with single rule', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
        disableRules: ['no-control-regex'],
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable no-control-regex */

      const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable` with multiple rules', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
        disableRules: ['no-control-regex', 'import/no-named-export'],
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable no-control-regex, import/no-named-export */

      const Minimum = z.string();
      "
    `)
  })

  test('Disable `.default()`', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        withoutDefaults: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z.string().min(1).max(10).regex(new RegExp("[a-z]+"));
      "
    `)
  })

  test('Enable `.descrive()`', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        withDesc: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z
        .string()
        .min(1)
        .max(10)
        .regex(new RegExp("[a-z]+"))
        .describe("bar")
        .default("foo");
      "
    `)
  })

  test('Wrap regex with anchors', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        withAnchors: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z
        .string()
        .min(1)
        .max(10)
        .regex(new RegExp("^[a-z]+$"))
        .default("foo");
      "
    `)
  })

  test('Use preset', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        parsers: {
          stringParser: 'format-regex-minmax',
        },
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z
        .string()
        .regex(new RegExp("[a-z]+"))
        .min(1)
        .max(10)
        .default("foo");
      "
    `)
  })

  test('Use external template', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        disableFormat: true,
        template: '__tests__/test.ts.ejs',
      })
    ).resolves.toMatchInlineSnapshot('"External Template"')
  })
})
