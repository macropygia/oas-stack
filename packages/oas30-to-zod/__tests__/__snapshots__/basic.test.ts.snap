// Vitest Snapshot v1

exports[`oasComponentsToZod (basic) > Complex regex 1`] = `
"const MailAddress = z
  .string()
  .regex(
    new RegExp(
      "(?:[a-z0-9!#$%&'*+/=?^_\`{|}~-]+(?:\\\\.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*|\\"(?:[\\\\x01-\\\\x08\\\\x0b\\\\x0c\\\\x0e-\\\\x1f\\\\x21\\\\x23-\\\\x5b\\\\x5d-\\\\x7f]|\\\\\\\\[\\\\x01-\\\\x09\\\\x0b\\\\x0c\\\\x0e-\\\\x7f])*\\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\\\x01-\\\\x08\\\\x0b\\\\x0c\\\\x0e-\\\\x1f\\\\x21-\\\\x5a\\\\x53-\\\\x7f]|\\\\\\\\[\\\\x01-\\\\x09\\\\x0b\\\\x0c\\\\x0e-\\\\x7f])+)\\\\])"
    )
  );
"
`;

exports[`oasComponentsToZod (basic) > Default 1`] = `
"import { z } from "zod";

const PrimitiveString = z.string();
const StringDefault = z.string().default("foo");
const StringDesc = z.string();
const StringMail = z.string().email();
const StringUrl = z.string().url();
const StringUuid = z.string().uuid();
const StringRange = z.string().min(10).max(20);
const StringPattern = z.string().regex(new RegExp("[a-z]+"));
const PrimitiveNumber = z.number();
const NumberMulitplier = z.number().multipleOf(2);
const NumberRange = z.number().gte(0).lte(100);
const NumberRangeExclusive = z.number().gt(0).lt(100);
const PrimitiveInteger = z.number().int();
const PrimitiveBoolean = z.boolean();
const NullablePrimitive = z.string().nullable();
const Array = z.array(z.string());
const ArrayRange = z.array(z.string()).min(1).max(10);
const ArrayLength = z.array(z.number()).length(3);
const NullableArray = z.array(z.number()).nullable();
const StringEnum = z.enum(["alpha", "beta"]);
const NumberEnum = z.union([z.literal(1.2), z.literal(34.5)]);
const IntegerEnum = z.union([z.literal(12), z.literal(345)]);
const NullableEnum = z.enum(["alpha", "beta"]).nullable();
const SingleEnum = z.literal(100);
const NonArrayEnum = z.literal(1000);
const TotallyObject = z.object({
  StringProp: z.string().optional(),
  NumberProp: z.number().optional(),
});
const MaybeObject = z.object({
  IntegerProp: z.number().int().optional(),
  StringArrayProp: z.array(z.string()).optional(),
});
const NullableObject = z
  .object({
    BooleanProp: z.boolean().optional(),
    StringEnumProp: z.enum(["delta", "gamma"]).optional(),
  })
  .nullable();
const AllOfObjects = z
  .object({ FooProp: z.string().optional(), BarProp: z.string().optional() })
  .merge(
    z.object({
      HogeProp: z.string().optional(),
      FugaProp: z.string().optional(),
    })
  );
const AllOfObjectRefs = TotallyObject.merge(MaybeObject);
const AllOfMixed = TotallyObject.merge(
  z.object({ FooProp: z.string().optional(), BarProp: z.string().optional() })
);
const AllOfInclNonObject = z.intersection(
  TotallyObject,
  z.intersection(PrimitiveString, MaybeObject)
);
const AllOfNone = z.undefined();
const AnyOfSingle = TotallyObject;
const AnyOfMultiple = z.undefined();
const AnyOfNone = z.undefined();
const OneOfSingle = TotallyObject;
const OneOfMultiple = z.union([TotallyObject, MaybeObject]);
const OneOfNone = z.undefined();

export const schemas = {
  PrimitiveString,
  StringDefault,
  StringDesc,
  StringMail,
  StringUrl,
  StringUuid,
  StringRange,
  StringPattern,
  PrimitiveNumber,
  NumberMulitplier,
  NumberRange,
  NumberRangeExclusive,
  PrimitiveInteger,
  PrimitiveBoolean,
  NullablePrimitive,
  Array,
  ArrayRange,
  ArrayLength,
  NullableArray,
  StringEnum,
  NumberEnum,
  IntegerEnum,
  NullableEnum,
  SingleEnum,
  NonArrayEnum,
  TotallyObject,
  MaybeObject,
  NullableObject,
  AllOfObjects,
  AllOfObjectRefs,
  AllOfMixed,
  AllOfInclNonObject,
  AllOfNone,
  AnyOfSingle,
  AnyOfMultiple,
  AnyOfNone,
  OneOfSingle,
  OneOfMultiple,
  OneOfNone,
};
"
`;

exports[`oasComponentsToZod (basic) > Disable format 1`] = `
"import { z } from 'zod';

const PrimitiveString = z.string();
const StringDefault = z.string().default("foo");
const StringDesc = z.string();
const StringMail = z.string().email();
const StringUrl = z.string().url();
const StringUuid = z.string().uuid();
const StringRange = z.string().min(10).max(20);
const StringPattern = z.string().regex(new RegExp("[a-z]+"));
const PrimitiveNumber = z.number();
const NumberMulitplier = z.number().multipleOf(2);
const NumberRange = z.number().gte(0).lte(100);
const NumberRangeExclusive = z.number().gt(0).lt(100);
const PrimitiveInteger = z.number().int();
const PrimitiveBoolean = z.boolean();
const NullablePrimitive = z.string().nullable();
const Array = z.array(z.string());
const ArrayRange = z.array(z.string()).min(1).max(10);
const ArrayLength = z.array(z.number()).length(3);
const NullableArray = z.array(z.number()).nullable();
const StringEnum = z.enum(["alpha","beta"]);
const NumberEnum = z.union([z.literal(1.2),z.literal(34.5)]);
const IntegerEnum = z.union([z.literal(12),z.literal(345)]);
const NullableEnum = z.enum(["alpha","beta"]).nullable();
const SingleEnum = z.literal(100);
const NonArrayEnum = z.literal(1000);
const TotallyObject = z.object({"StringProp":z.string().optional(),"NumberProp":z.number().optional()});
const MaybeObject = z.object({"IntegerProp":z.number().int().optional(),"StringArrayProp":z.array(z.string()).optional()});
const NullableObject = z.object({"BooleanProp":z.boolean().optional(),"StringEnumProp":z.enum(["delta","gamma"]).optional()}).nullable();
const AllOfObjects = z.object({"FooProp":z.string().optional(),"BarProp":z.string().optional()}).merge(z.object({"HogeProp":z.string().optional(),"FugaProp":z.string().optional()}));
const AllOfObjectRefs = TotallyObject.merge(MaybeObject);
const AllOfMixed = TotallyObject.merge(z.object({"FooProp":z.string().optional(),"BarProp":z.string().optional()}));
const AllOfInclNonObject = z.intersection(TotallyObject,z.intersection(PrimitiveString,MaybeObject));
const AllOfNone = z.undefined();
const AnyOfSingle = TotallyObject;
const AnyOfMultiple = z.undefined();
const AnyOfNone = z.undefined();
const OneOfSingle = TotallyObject;
const OneOfMultiple = z.union([TotallyObject,MaybeObject]);
const OneOfNone = z.undefined();

export const schemas = {
  PrimitiveString,
  StringDefault,
  StringDesc,
  StringMail,
  StringUrl,
  StringUuid,
  StringRange,
  StringPattern,
  PrimitiveNumber,
  NumberMulitplier,
  NumberRange,
  NumberRangeExclusive,
  PrimitiveInteger,
  PrimitiveBoolean,
  NullablePrimitive,
  Array,
  ArrayRange,
  ArrayLength,
  NullableArray,
  StringEnum,
  NumberEnum,
  IntegerEnum,
  NullableEnum,
  SingleEnum,
  NonArrayEnum,
  TotallyObject,
  MaybeObject,
  NullableObject,
  AllOfObjects,
  AllOfObjectRefs,
  AllOfMixed,
  AllOfInclNonObject,
  AllOfNone,
  AnyOfSingle,
  AnyOfMultiple,
  AnyOfNone,
  OneOfSingle,
  OneOfMultiple,
  OneOfNone
};
"
`;

exports[`oasComponentsToZod (basic) > Enable dereference 1`] = `
"const Comp1 = z.object({
  Prop1: z.string().optional(),
  Prop2: z.number().optional(),
});
const Comp2 = z.object({
  Prop1: z.string().optional(),
  Prop2: z.number().optional(),
});
"
`;

exports[`oasComponentsToZod (basic) > Wrap regex with anchors 1`] = `
"const MailAddress = z
  .string()
  .regex(
    new RegExp(
      "^(?:[a-z0-9!#$%&'*+/=?^_\`{|}~-]+(?:\\\\.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*|\\"(?:[\\\\x01-\\\\x08\\\\x0b\\\\x0c\\\\x0e-\\\\x1f\\\\x21\\\\x23-\\\\x5b\\\\x5d-\\\\x7f]|\\\\\\\\[\\\\x01-\\\\x09\\\\x0b\\\\x0c\\\\x0e-\\\\x7f])*\\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\\\x01-\\\\x08\\\\x0b\\\\x0c\\\\x0e-\\\\x1f\\\\x21-\\\\x5a\\\\x53-\\\\x7f]|\\\\\\\\[\\\\x01-\\\\x09\\\\x0b\\\\x0c\\\\x0e-\\\\x7f])+)\\\\])$"
    )
  );
"
`;
